spacy_curated_transformers/models/listeners.py

```diff
@@ -435,75 +528,244 @@ def __init__(
-            return [model.ops.alloc2f(len(doc), width) for doc in docs], lambda dY: []
+            return Y, backprop
+        else:
+            if TransformerListener.use_doc_annotations_for_prediction(model):
+                width = model.get_dim("nO")
+
+                no_trf_data = [doc._.trf_data is None for doc in docs]
+                if any(no_trf_data):
+                    assert all(no_trf_data)
+                    return [
+                        model.ops.alloc2f(len(doc), width) for doc in docs
+                    ], lambda dY: []
+
+                if any(doc._.trf_data.last_layer_only for doc in docs):
+                    raise ValueError(
+                        Errors.E012.format(listener_name="ScalarWeightingListener")
+                    )
+
+                Y_weighting = weighting.predict(
+                    [doc._.trf_data.all_outputs for doc in docs]
+                )
+                Y = pooling.predict(Y_weighting)
 
-        if any(doc._.trf_data.last_layer_only for doc in docs):
-            raise ValueError(
-                Errors.E012.format(listener_name="ScalarWeightingListener")
-            )
+                return Y, lambda dX: []
+            else:
+                assert _outputs is not None
+                TransformerListener.verify_inputs(model, docs)
+                Y_weighting = weighting.predict(_outputs.all_outputs)
+                outputs: Tuple[List[Floats2d], Callable[[Any], Any]] = (
+                    pooling.predict(Y_weighting),
+                    lambda dY: [],
+                )
+                TransformerListener.clear_state(model)
+                return outputs
 
-        Y_weigthing = weighting.predict([doc._.trf_data.all_outputs for doc in docs])
-        Y = pooling.predict(Y_weigthing)
 
-        return Y, lambda dX: []
+class WrappedTransformerAndListener(WrappedTransformerAndListenerModelT):
+    """Wraps a transformer model and a compatible listener. Exclusively used when replacing
+    listeners of a shared transformer pipeline."""
+
+    name = "wrapped_transformer_and_listener"
+
+    def __init__(
+        self,
+        transformer: TransformerModelT,
+        listener: TransformerListenerModelT,
+        frozen: bool = False,
+    ) -> None:
+        """
+        transformer (TransformerModelT):
+            Transformer model to wrap.
+        listener (TransformerListenerModelT):
+            Listener model to wrap.
+        frozen (bool):
+            If the transformer is frozen.
+        """
+        Model.__init__(
+            self,
+            name=self.name,
+            init=wrapped_transformer_and_listener_init,
+            forward=wrapped_transformer_and_listener_forward,
+            dims={"nO": listener.get_dim("nO")},
+            layers=[transformer, listener],
+        )
+
+        # Ensure that the transformer returns the required outputs.
+        transformer.attrs["_all_layer_outputs"] = listener.name in (
```

Comment:
Maybe we should add an attribute to the model for this? As it is, it is not possible for a third party to make their own multi-layer listener

Reply:
Good idea. Added.